File: examples/pascal/constructs/short_circuit.pas
// Ported from examples/lumen/constructs/short_circuit.lm by scripts/port_examples.py; edit the Lumen original, not this file.
var result: boolean;
var x: integer;

function is_even(x: integer): boolean;
begin
    writeln('Checking if ');
    writeln(x);
    writeln(' is even');
    is_even := x mod 2 = 0;
end;

function is_positive(x: integer): boolean;
begin
    writeln('Checking if ');
    writeln(x);
    writeln(' is positive');
    is_positive := x > 0;
end;

procedure safe_check(value: integer);
begin
    if (value <> nil) and (value > 10) then begin
        writeln('Value is not null and greater than 10');
    end else begin
        writeln('Value is null or not greater than 10');
    end;
end;

begin
    writeln('Test: Short-Circuit Evaluation');
    writeln('false and is_even(10):');
    result := false and is_even(10);
    writeln(result);
    writeln('true and is_even(10):');
    result := true and is_even(10);
    writeln(result);
    writeln('true or is_positive(5):');
    result := true or is_positive(5);
    writeln(result);
    writeln('false or is_positive(5):');
    result := false or is_positive(5);
    writeln(result);
    writeln('Testing division by zero avoidance:');
    x := 0;
    if (x <> 0) and (10 / x > 5) then begin
        writeln('Result is greater than 5');
    end else begin
        writeln('x is zero or result is not greater than 5');
    end;
    safe_check(15);
    safe_check(5);
end.

File: examples/javascript/constructs/short_circuit.js
// Ported from examples/lumen/constructs/short_circuit.lm by scripts/port_examples.py; edit the Lumen original, not this file.
function is_even(x) {
    console.log("Checking if ");
    console.log(x);
    console.log(" is even");
    return x % 2 === 0;
}

function is_positive(x) {
    console.log("Checking if ");
    console.log(x);
    console.log(" is positive");
    return x > 0;
}

console.log("Test: Short-Circuit Evaluation");
console.log("false and is_even(10):");
let result = false && is_even(10);
console.log(result);
console.log("true and is_even(10):");
result = true && is_even(10);
console.log(result);
console.log("true or is_positive(5):");
result = true || is_positive(5);
console.log(result);
console.log("false or is_positive(5):");
result = false || is_positive(5);
console.log(result);
console.log("Testing division by zero avoidance:");
const x = 0;
if (x !== 0 && 10 / x > 5) {
    console.log("Result is greater than 5");
} else {
    console.log("x is zero or result is not greater than 5");
}
function safe_check(value) {
    if (value !== null && value > 10) {
        console.log("Value is not null and greater than 10");
    } else {
        console.log("Value is null or not greater than 10");
    }
}

safe_check(15);
safe_check(5);
